src/components/CodeAssignmentModal/index.jsx

```diff
@@ -0,0 +1,338 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Field, reduxForm, SubmissionError } from 'redux-form';
+import { Button, Icon, Modal } from '@edx/paragon';
+import isEmail from 'validator/lib/isEmail';
+
+import H3 from '../H3';
+import TextAreaAutoSize from '../TextAreaAutoSize';
+import StatusAlert from '../StatusAlert';
+import BulkAssignFields from './BulkAssignFields';
+import IndividualAssignFields from './IndividualAssignFields';
+
+import EcommerceApiService from '../../data/services/EcommerceApiService';
+
+import emailTemplate from './emailTemplate';
+
+import './CodeAssignmentModal.scss';
+
+class CodeAssignmentModal extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.errorMessageRef = React.createRef();
+    this.modalRef = React.createRef();
+
+    this.validateFormData = this.validateFormData.bind(this);
+    this.handleModalSubmit = this.handleModalSubmit.bind(this);
+  }
+
+  componentDidMount() {
+    const { current: { firstFocusableElement } } = this.modalRef;
+
+    if (firstFocusableElement) {
+      firstFocusableElement.focus();
+    }
+  }
+
+  componentDidUpdate(prevProps) {
+    const {
+      submitFailed,
+      submitSucceeded,
+      onClose,
+      error,
+    } = this.props;
+
+    const errorMessageRef = this.errorMessageRef && this.errorMessageRef.current;
+
+    if (submitSucceeded && submitSucceeded !== prevProps.submitSucceeded) {
+      onClose();
+    }
+
+    if (submitFailed && error && error.length > 0 && error !== prevProps.error && errorMessageRef) {
+      // When there is an error, focus on the error message status alert
+      errorMessageRef.focus();
+    }
+  }
+
+  validateBulkAssign(formData) {
+    const { data: { unassignedCodes } } = this.props;
```

Comment:
i'm not quite able to tell where unassignedCodes comes from, but does it mean codes or code usages? there could be just one unassigned code, but it could have enough usages left for assignment, and i can't tell if the logic below would reflect that.

Reply:
`unassignedCodes` is passed to the `CodeAssignmentModal` as a [key within the `data` object prop](https://github.com/edx/edx-portal/pull/149/files#diff-6a1729462bebbd0c4ca0ec8b6fe2d0daR204).

The actual source of the `unassignedCodes` value is from the `num_unassigned` field from the coupon overview API endpoint ([see example response here](https://ecommerce-business.sandbox.edx.org/api/v2/enterprise/coupons/1b542b06d35345e698679579d9e12745/overview/))., e.g.:

```json
"results": [
    {
      "end_date": "2019-01-26T00:00:00Z",
      "has_error": false,
      "id": 418,
      "max_uses": null,
      "num_codes": 1,
      "num_unassigned": 0,
      "num_uses": 0,
      "start_date": "2019-01-05T00:00:00Z",
      "title": "123456",
      "usage_limitation": "Single use"
    },
],
```